bindings/java/src/main/java/io/github/treesitter/jtreesitter/java/TreeSitterJava.java

```diff
@@ -0,0 +1,38 @@
+package io.github.treesitter.jtreesitter.java;
+
+import java.lang.foreign.*;
+import java.lang.invoke.MethodHandle;
+
+/** Java grammar for tree-sitter */
+public final class TreeSitterJava {
+    private static final ValueLayout VOID_PTR =
+            ValueLayout.ADDRESS.withTargetLayout(MemoryLayout.sequenceLayout(Long.MAX_VALUE, ValueLayout.JAVA_BYTE));
+    private static final FunctionDescriptor FUNC_DESC = FunctionDescriptor.of(VOID_PTR);
+    private static final Linker LINKER = Linker.nativeLinker();
+    private static final TreeSitterJava INSTANCE = new TreeSitterJava();
+
+    private final Arena arena = Arena.ofAuto();
+    private final String library = System.mapLibraryName("tree-sitter-java");
+    private final SymbolLookup symbols =
+            SymbolLookup.libraryLookup(library, arena).or(SymbolLookup.loaderLookup());
+
+    /** Get the tree-sitter {@code Language} for this grammar */
+    public static MemorySegment language() {
```

Comment:
Would it make sense to have two variants?
- `TreeSitterJava.language()`
- `TreeSitterJava.language(SymbolLookup)`

The former could then simply delegate to the latter, and use your existing logic here with `System.mapLibraryName(...)`.

It can be useful if users can provide the `SymbolLookup` themselves:
- It allows them to control where the library is loaded from, e.g. they can use a temp file
- They can unload the library by calling `Arena#close` (see [`SymbolLookup#libraryLookup` documentation](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/SymbolLookup.html#libraryLookup(java.nio.file.Path,java.lang.foreign.Arena)))
On Linux that might not matter that much, but on Windows you cannot delete the library file while it is still loaded which prevents clean-up of a temporarily extracted library file

Reply:
Libraries can be loaded from arbitrary locations by setting `java.library.path`.
That's what `or(SymbolLookup.loaderLookup())` is for.

Reply:
> Libraries can be loaded from arbitrary locations by setting `java.library.path`.

`SymbolLookup.libraryLookup` does not seem to consider `java.library.path`, see [JDK-8311090](https://bugs.openjdk.org/browse/JDK-8311090).

---

> That's what `or(SymbolLookup.loaderLookup())` is for.

As mentioned in my other comment https://github.com/tree-sitter/tree-sitter-java/pull/182#discussion_r1740185719, I think that does not really work. If `SymbolLookup.libraryLookup` can't find the library it already throws an exception: I am talking about the `libraryLookup` call already, not later the `SymbolLookup#find` call.

So the user does not actually have a choice to provide the library though different means (e.g. `System#load` which would be considered by `SymbolLookup.loaderLookup()`).

---

> The library file should not be deleted _while it's loaded_.

That is my point: If you allow the user to provide their own `SymbolLookup`, they can then later call `Arena#close` to unload the library, see [documentation](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/SymbolLookup.html#libraryLookup(java.nio.file.Path,java.lang.foreign.Arena)):

> For instance, if the provided arena is a confined arena, the library associated with the returned lookup will be unloaded when the provided confined arena is closed.

Otherwise it depends on garbage collection when or even if at all the library is unloaded.

Reply:
So if users cannot provide their own `SymbolLookup`, then they are a bit restricted:
- They have to adhere to the OS-specific locations of library lookup: On Windows this is `PATH` (?), on Linux `LD_LIBRARY_PATH` (which does not include the working dir by default?)
- They cannot control if / when the library is unloaded